src/components/asset-list/RecyclerAssetList.js

```diff
@@ -277,50 +271,76 @@ class RecyclerAssetList extends Component {
               }
             }
             if (
               sections[balancesIndex].data[lastBalanceIndex - 2]
                 .smallBalancesContainer
             ) {
-              return ViewTypes.COIN_DIVIDER;
+              return {
+                height: ViewTypes.COIN_DIVIDER.calculateHeight(),
+                index: ViewTypes.COIN_DIVIDER.index,
+                visibleDuringCoinEdit:
+                  ViewTypes.COIN_DIVIDER.visibleDuringCoinEdit,
+              };
             }
           }
           if (index === lastBalanceIndex - 1) {
             if (
+              sections[balancesIndex].data[lastBalanceIndex - 2] &&
               sections[balancesIndex].data[lastBalanceIndex - 2]
                 .smallBalancesContainer
             ) {
               smallBalancesIndex = index - 1;
-              return ViewTypes.COIN_SMALL_BALANCES;
+              return {
+                height: ViewTypes.COIN_SMALL_BALANCES.calculateHeight({
+                  isCoinListEdited: this.props.isCoinListEdited,
+                  isOpen: this.props.openSmallBalances,
+                  smallBalancesLength:
+                    sections[balancesIndex].data[smallBalancesIndex].assets
+                      .length,
+                }),
+                index: ViewTypes.COIN_SMALL_BALANCES.index,
+                visibleDuringCoinEdit:
+                  ViewTypes.COIN_SMALL_BALANCES.visibleDuringCoinEdit,
+              };
             }
           }
           if (index === lastBalanceIndex) {
             if (
-              sections[balancesIndex].data[lastBalanceIndex - 1]
-                .smallBalancesContainer
-            ) {
-              smallBalancesIndex = index - 1;
-              return ViewTypes.COIN_SMALL_BALANCES;
-            } else if (
               sections[balancesIndex].data[lastBalanceIndex - 1]
                 .savingsContainer
             ) {
-              savingsIndex = index - 1;
-              return ViewTypes.COIN_SAVINGS;
+              return {
+                height: ViewTypes.COIN_SAVINGS.calculateHeight({
+                  amountOfRows:
+                    sections[balancesIndex].data[index - 1].assets.length,
+                  isOpen: this.props.openSavings,
+                }),
+                index: ViewTypes.COIN_SAVINGS.index,
+              };
             }
             this.lastAssetIndex = index;
-            return ViewTypes.COIN_ROW_LAST;
           }
           const firstBalanceIndex = headersIndices[balancesIndex] + 1;
-          if (
-            index === firstBalanceIndex &&
-            !sections[balancesIndex].data[firstBalanceIndex - 1]
-              .smallBalancesContainer
-          ) {
-            return ViewTypes.COIN_ROW_FIRST;
-          }
+
+          return {
+            height: ViewTypes.COIN_ROW.calculateHeight({
+              areSmallCollectibles: this.state.areSmallCollectibles,
+              isFirst:
```

Comment:
can we pull out the isFirst logic on line 331 and line 338 as a separate const called isFirstCoinRow or something

Reply:
Great idea 🎉
Applied